week4/assignment/auth-practice/src/hooks/useJoin.ts

```diff
@@ -0,0 +1,149 @@
+import { postJoin } from '@apis/member';
+import { formatPhoneNumber, validatePassword } from '@utils/utils';
+import React, { useRef, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+const useJoin = () => {
+  const navigate = useNavigate();
+
+  // input focus를 위한 ref
+  const idRef = useRef<HTMLInputElement>(null);
+  const passwordRef = useRef<HTMLInputElement>(null);
+  const nicknameRef = useRef<HTMLInputElement>(null);
+  const phoneNumberRef = useRef<HTMLInputElement>(null);
+
+  // id, password, nickname, phoneNumber를 관리하는 state
+  const [id, setId] = useState('');
+  const [password, setPassword] = useState('');
+  const [nickname, setNickname] = useState('');
+  const [phoneNumber, setPhoneNumber] = useState('');
+
+  // id, password, nickname, phoneNumber 에러 여부를 관리하는 state
+  const [idError, setIdError] = useState(false);
+  const [passwordError, setPasswordError] = useState(false);
+  const [nicknameError, setNicknameError] = useState(false);
+  const [phoneNumberError, setPhoneNumberError] = useState(false);
+
+  // id input이 변경될 때 호출되는 함수
+  const handleChangeId = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setId(e.target.value);
+  };
+  // password input이 변경될 때 호출되는 함수
+  const handleChangePassword = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+  };
+  // nickname input이 변경될 때 호출되는 함수
+  const handleChangeNickname = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setNickname(e.target.value);
+  };
+  // phoneNumber input이 변경될 때 호출되는 함수
+  const handleChangePhoneNumber = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPhoneNumber(formatPhoneNumber(e.target.value));
+  };
+
+  // 회원가입 버튼 클릭 시 호출되는 함수
+  // id, password, nickname, phoneNumber가 유효한지 검사하는 함수
+  const validateJoin = () => {
+    // id, password, nickname, phoneNumber 에러 여부 초기화
+    setIdError(false);
+    setPasswordError(false);
+    setNicknameError(false);
+    setPhoneNumberError(false);
+
+    // id가 입력되지 않았을 경우 alert 출력 후 focus
+    if (!id) {
+      alert('아이디를 입력해주세요.');
+      if (idRef.current) {
+        idRef.current.focus();
+      }
```

Comment:
p5: TS의 단언 연산자를 아시나요...? (!) 연산자를 사용하면 다음과 같이 사용할 수 있어요!
```suggestion
      idRef.current!.focus();
```
https://developer-talk.tistory.com/191

Reply:
!연산자는 null 이 아님을 나타내는 것으로 알고있긴 했는데, 이렇게 사용해서 코드를 좀더 간결하게 작성할 수 있네요!